Improve performance of InCodeGenerator and ExpressionInterpreter

raunaqmorarka opened the pull request:
Going through MethodHandle for hashCode and equals operations is much slower than direct method access.
Using compileSingleAccessMethod to generate the required interface from method handles results in around 50% improvement on BenchmarkInCodeGenerator for long and double.

losipiuk commented on core/trino-main/src/main/java/io/trino/type/SingleAccessMethodCompiler.java:
```diff
@@ -39,7 +39,7 @@
 import static java.lang.invoke.MethodType.methodType;
 
-final class SingleAccessMethodCompiler
+public final class SingleAccessMethodCompiler
```
nit: move to `io.trino.util`

raunaqmorarka replied:
added a commit with that change

losipiuk commented on core/trino-main/src/main/java/io/trino/util/FastutilSetHelper.java:
```diff
@@ -88,97 +96,112 @@ public static boolean in(Object objectValue, ObjectOpenCustomHashSet<?> set)
+
+        public LongStrategy(MethodHandle hashCodeHandle, MethodHandle equalsHandle, Type type)
+        {
+            requireNonNull(hashCodeHandle, "hashCodeHandle is null");
```
nit: rnn on type

losipiuk commented on core/trino-main/src/main/java/io/trino/util/FastutilSetHelper.java:
```diff
@@ -203,16 +219,93 @@ public boolean equals(Object a, Object b)
+        }
+    }
+
+    static class MethodGenerator
```
private?

raunaqmorarka replied:
done

losipiuk reviewed (approved):
Nice :)

losipiuk commented on core/trino-main/src/main/java/io/trino/util/FastutilSetHelper.java:
```diff
@@ -203,16 +229,93 @@ public boolean equals(Object a, Object b)
+                this.methodHandle = requireNonNull(methodHandle, "methodHandle is null");
+            }
+
+            public synchronized T get()
+            {
+                if (operator != null) {
+                    return operator;
+                }
+                String suggestedClassName = singleMethodInterface.getSimpleName() + "_" + type.getDisplayName();
+                operator = compileSingleAccessMethod(suggestedClassName, singleMethodInterface, methodHandle);
+                // Free up memory by removing references to the objects that we won't need anymore
+                type = null;
+                singleMethodInterface = null;
+                methodHandle = null;
+                return operator;
```
Actually a question. Is there a reason to do this part lazyily. Why not eagerly generate `T` and store it in cache?

raunaqmorarka replied:
For this I was trying to stay close to `BlockTypeOperators#getBlockOperator` as that is doing very similar thing. I don't see any drawback in the simpler eager generation approach though. Either way we should end up holding lock on only one entry.
@dain please chime in if there is a specific reason to prefer the lazy loading from `BlockTypeOperators#getBlockOperator`.

raunaqmorarka replied:
I tried to dig a bit into the granularity of locking used by guava Cache. I get the impression that it locks a segment of the Map only to insert entry (which we would encounter either way), but for the load of the entry it takes lock only on the newly inserted entry which should be be similar to what we did here.
But I'm happy to stick to the current way for the sake of consistency and simpler exception handling.

dain replied:
IIRC Guava uses striped lockes.  Typically a system like this uses 1 lock per core.  I think the bigger win is a simpler stack when things go wrong.  Also, in the case of a failure, you get one total object instead of one object per request to the cache

martint commented:
Why is the first commit a “revert”? Even if the mechanics of the change were a “git revert”, rephrase the message in terms of the logical result it achieves: “move xxx to yyy”

raunaqmorarka commented:
> Why is the first commit a “revert”? Even if the mechanics of the change were a “git revert”, rephrase the message in terms of the logical result it achieves: “move xxx to yyy”

I've updated the first commit title to `Move FastutilSetHelper to trino-main` now. This allows easy usage of SingleAccessMethodCompiler.java which is in trino-main.
The original use case for which I had moved FastutilSetHelper to trino-plugin-toolkit turned out to have another better alternative, so the original change is no longer necessary.

sopel39 commented:
nice!
